grudge/geometry/metrics.py

```diff
@@ -0,0 +1,712 @@
+            (E, F), (_, G) = form1      # noqa: N806
+            inv_form1 = 1.0 / (E * G - F * F) * np.stack(
+                [make_obj_array([G, -F]),
+                 make_obj_array([-F, E])]
+            )
+        else:
+            raise ValueError("%dD surfaces not supported" % dim)
+
+    return inv_form1
+
+
+def inverse_metric_derivative(actx, dcoll, rst_axis, xyz_axis, dd):
+    r"""Computes the inverse metric derivative of the physical
+    coordinate enumerated by *xyz_axis* with respect to the
+    reference axis *rst_axis*.
+
+    :arg actx: an :class:`~arraycontext.context.ArrayContext`.
+    :arg dcoll: a :class:`grudge.discretization.DiscretizationCollection`.
+    :arg rst_axis: an integer denoting the reference coordinate axis.
+    :arg xyz_axis: an integer denoting the physical coordinate axis.
+    :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+        Defaults to the base volume discretization.
+    :returns: a :class:`~meshmode.dof_array.DOFArray` containing the
+        inverse metric derivative at each nodal coordinate.
+    """
+
+    dim = dcoll.dim
+    if dim != dcoll.ambient_dim:
+        raise ValueError(
+            "Not clear what inverse_metric_derivative means if "
+            "the derivative matrix is not square!"
+        )
+
+    par_vecs = [forward_metric_derivative_mv(actx, dcoll, rst, dd)
+                for rst in range(dim)]
+
+    # Yay Cramer's rule!
+    from functools import reduce, partial
+    from operator import xor as outerprod_op
+    outerprod = partial(reduce, outerprod_op)
+
+    def outprod_with_unit(i, at):
+        unit_vec = np.zeros(dim)
+        unit_vec[i] = 1
+
+        vecs = par_vecs[:]
+        vecs[at] = MultiVector(unit_vec)
+
+        return outerprod(vecs)
+
+    volume_pseudoscalar_inv = outerprod(
+        forward_metric_derivative_mv(actx, dcoll, rst_axis, dd)
+        for rst_axis in range(dim)
+    ).inv()
+
+    result = (outprod_with_unit(xyz_axis, rst_axis)
+              * volume_pseudoscalar_inv).as_scalar()
+
+    return result
+
+
+@memoize_on_first_arg
+def inverse_surface_metric_derivative(actx, dcoll, rst_axis, xyz_axis, dd=None):
+    r"""Computes the inverse surface metric derivative of the physical
+    coordinate enumerated by *xyz_axis* with respect to the
+    reference axis *rst_axis*. These geometric terms are used in the
+    transformation of physical gradients.
+
+    :arg actx: an :class:`~arraycontext.context.ArrayContext`.
+    :arg dcoll: a :class:`grudge.discretization.DiscretizationCollection`.
+    :arg rst_axis: an integer denoting the reference coordinate axis.
+    :arg xyz_axis: an integer denoting the physical coordinate axis.
+    :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+        Defaults to the base volume discretization.
+    :returns: a :class:`~meshmode.dof_array.DOFArray` containing the
+        inverse metric derivative at each nodal coordinate.
```

Comment:
Yep. It's returning a `DOFArray` because `forward_metric_nth_derivative` (which most of everything in `metrics.py` eventually calls) returns a `DOFArray`. So `ndarray` and `MultiVector` containing `DOFArrays` are being passed around.